Ical.Net/DataTypes/CalDateTime.cs

```
@@ -486,126 +509,139 @@ TzId is null
     {
-        var dt = Copy<CalDateTime>();
-        if (!dt.HasTime && hours % 24 > 0)
+        var copy = Copy<CalDateTime>();
+        var newValue = copy.Value.AddHours(hours);
+        if (!copy.HasTime && (hours % 24 == 0))
         {
-            dt.HasTime = true;
+            copy._dateOnly = DateOnly.FromDateTime(newValue);
         }
-        dt.Value = Value.AddHours(hours);
-        return dt;
+        else
+        {
+            copy._dateOnly = DateOnly.FromDateTime(newValue);
+            copy._timeOnly = RoundTimeToNearestSecond(newValue);
+        }
+        return copy;
     }
 
     /// <inheritdoc cref="DateTime.AddMinutes"/>
     /// <remarks>
-    /// This will also set <seealso cref="HasTime"/> to <see langword="true"/>,
+    /// This will also add a <see cref="IDateTime.Time"/> part that did not exist before the operation,
     /// if the minutes are not a multiple of 1440.
     /// </remarks>
     public IDateTime AddMinutes(int minutes)
     {
-        var dt = Copy<CalDateTime>();
-        if (!dt.HasTime && minutes % 1440 > 0)
+        var copy = Copy<CalDateTime>();
+        var newValue = copy.Value.AddMinutes(minutes);
+        if (!copy.HasTime && (minutes % 1440 == 0))
         {
-            dt.HasTime = true;
+            copy._dateOnly = DateOnly.FromDateTime(newValue);
         }
-        dt.Value = Value.AddMinutes(minutes);
-        return dt;
+        else
+        {
+            copy._dateOnly = DateOnly.FromDateTime(newValue);
+            copy._timeOnly = RoundTimeToNearestSecond(newValue);
+        }
+        return copy;
     }
 
     /// <inheritdoc cref="DateTime.AddSeconds"/>
     /// <remarks>
-    /// This will also set <seealso cref="HasTime"/> to <see langword="true"/>,
+    /// This will also add a <see cref="IDateTime.Time"/> part that did not exist before the operation,
     /// if the seconds are not a multiple of 86400.
     /// </remarks>
     public IDateTime AddSeconds(int seconds)
```

I think we should raise exceptions if `Add[sec|min|hour]` is called. If people want to convert a DATE to a DATE-TIME, then they should do this explicitly and also specify a tzid (or intentionally none).

---

Happy to hear your suggestion, as this was my first conclusion, too. It's one of "guess the users intention".
Let's throw an `InvalidOperationException`. Will modify.